app/kuma-prometheus-sd/pkg/discovery/xds/xds.go

```diff
@@ -0,0 +1,171 @@
+	mads_client "github.com/Kong/kuma/pkg/mads/client"
+)
+
+type DiscoveryConfig struct {
+	ServerURL  string
+	ClientName string
+}
+
+type discoverer struct {
+	log           logr.Logger
+	config        DiscoveryConfig
+	oldSourceList map[string]bool
+}
+
+func NewDiscoverer(config DiscoveryConfig, log logr.Logger) (*discoverer, error) {
+	cd := &discoverer{
+		log:           log,
+		config:        config,
+		oldSourceList: make(map[string]bool),
+	}
+	return cd, nil
+}
+
+// Run implements discovery.Discoverer interface.
+func (d *discoverer) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
+	// notice that Prometheus discovery.Discoverer abstraction doesn't allow failures,
+	// so we must ensure that xDS client is up-and-running all the time.
+	for streamID := uint64(1); ; streamID++ {
+		errCh := make(chan error, 1)
+		go func(errCh chan<- error) {
+			defer close(errCh)
+			// recover from a panic
+			defer func() {
+				if e := recover(); e != nil {
+					if err, ok := e.(error); ok {
+						errCh <- err
+					} else {
+						errCh <- errors.Errorf("%v", e)
+					}
+				}
+			}()
+			errCh <- d.stream(ctx, streamID, ch)
+		}(errCh)
+		select {
+		case <-ctx.Done():
+			d.log.Info("done")
+			break
+		case err := <-errCh:
+			if err != nil {
+				d.log.WithValues("streamID", streamID).Error(err, "xDS stream terminated with an error")
+			}
+		}
+	}
+}
+
+func (d *discoverer) stream(ctx context.Context, streamID uint64, ch chan<- []*targetgroup.Group) error {
+	log := d.log.WithValues("streamID", streamID)
+
+	log.Info("creating a gRPC client for Monitoring Assignment Discovery Service (MADS) server ...")
+	client, err := mads_client.New(d.config.ServerURL)
+	if err != nil {
+		return errors.Wrap(err, "failed to connect to gRPC server")
+	}
+	defer client.Close()
+
+	log.Info("starting an xDS stream ...")
+	stream, err := client.StartStream()
+	if err != nil {
+		return errors.Wrap(err, "failed to start an xDS stream")
+	}
+	defer stream.Close()
+
+	log.Info("sending first discovery request on a new xDS stream ...")
+	err = stream.RequestAssignments(d.config.ClientName)
+	if err != nil {
+		return errors.Wrap(err, "failed to send a discovery request")
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+
+		log.Info("waiting for a discovery response ...")
+		assignments, err := stream.WaitForAssignments()
+		if err != nil {
+			return errors.Wrap(err, "failed to receive a discovery response")
+		}
+		log.Info("received monitoring assignments", "len", len(assignments))
+		log.V(1).Info("received monitoring assignments", "assignments", assignments)
+
+		tgs, newSourceList := d.convertAll(assignments)
+		ch <- tgs
+		d.oldSourceList = newSourceList
+
+		if err := stream.ACK(); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return errors.Wrap(err, "failed to ACK a discovery response")
+		}
+	}
+
+	log.Info("closing an xDS stream ...")
+	if err := stream.Close(); err != nil {
```

Comment:
I see that `Close()` was already called in defer. Why call it two times?

Reply:
reworked